pkg/envoy/xds/server_e2e_test.go

```diff
@@ -0,0 +1,474 @@
+	}
+	typeURL, ok := params[4].(string)
+	if !ok {
+		return false, "TypeURL must be a string"
+	}
+
+	error = ""
+
+	result = response.VersionInfo == versionInfo &&
+		len(response.Resources) == len(resources) &&
+		response.Canary == canary &&
+		response.TypeUrl == typeURL
+
+	if result && len(resources) > 0 {
+		// Convert the resources into Any protocol buffer messages, which is
+		// the type of Resources in the response, so that we can compare them.
+		resourcesAny := make([]*any.Any, 0, len(resources))
+		for _, res := range resources {
+			data, err := proto.Marshal(res)
+			if err != nil {
+				return false, fmt.Sprintf("error marshalling protocol buffer %v", res)
+			}
+			resourcesAny = append(resourcesAny,
+				&any.Any{
+					TypeUrl: typeURL,
+					Value:   data,
+				})
+		}
+		// Sort both lists.
+		sort.Slice(response.Resources, func(i, j int) bool {
+			return response.Resources[i].String() < response.Resources[j].String()
+		})
+		sort.Slice(resourcesAny, func(i, j int) bool {
+			return resourcesAny[i].String() < resourcesAny[j].String()
+		})
+		result = reflect.DeepEqual(response.Resources, resourcesAny)
+	}
+
+	return
+}
+
+// ResponseMatches checks that a DiscoveryResponse's fields match the given
+// parameters.
+var ResponseMatches Checker = &ResponseMatchesChecker{
+	&CheckerInfo{Name: "ResponseMatches", Params: []string{
+		"response", "VersionInfo", "Resources", "Canary", "TypeUrl"}},
+}
+
+var resources = []*api.RouteConfiguration{
+	{Name: "resource0"},
+	{Name: "resource1"},
+	{Name: "resource2"},
+}
+
+func (s *ServerSuite) TestRequestAllResources(c *C) {
+	typeURL := "type.googleapis.com/envoy.api.v2.DummyConfiguration"
+
+	var err error
+	var req *api.DiscoveryRequest
+	var resp *api.DiscoveryResponse
+	var v uint64
+	var mod bool
+
+	ctx, cancel := context.WithTimeout(context.Background(), TestTimeout)
+	defer cancel()
+
+	cache := NewCache()
+
+	streamCtx, closeStream := context.WithCancel(ctx)
+	stream := NewMockStream(streamCtx, 1, 1, StreamTimeout, StreamTimeout)
+	defer stream.Close()
+
+	server := NewServer(map[string]ObservableResourceSource{
+		typeURL: cache,
+	}, TestTimeout)
+
+	streamDone := make(chan struct{})
+
+	// Run the server's stream handler concurrently.
+	go func() {
+		err := server.HandleRequestStream(ctx, stream, AnyTypeURL)
+		close(streamDone)
+		c.Check(err, IsNil)
+	}()
+
+	// Request all resources.
+	req = &api.DiscoveryRequest{
+		TypeUrl:       typeURL,
+		VersionInfo:   "",
+		Node:          nil, // Ignored by server.
+		ResourceNames: nil,
+		ResponseNonce: "",
+	}
+	err = stream.SendRequest(req)
+	c.Assert(err, IsNil)
+
+	// Expecting an empty response.
+	resp, err = stream.RecvResponse()
+	c.Assert(err, IsNil)
+	c.Assert(resp, ResponseMatches, "0", nil, false, typeURL)
+	c.Assert(resp.Nonce, Not(Equals), "")
+
+	// Request the next version of resources.
+	req.VersionInfo = resp.VersionInfo // ACK the received version.
+	req.ResponseNonce = resp.Nonce
```

Comment:
Maybe add a test where the the client sends a stale nonce in the request to exercise that code path on the server and verify that there is no response, and that the next request with the correct nonce gets a response?

Reply:
Done.

Reply:
Writing this test allowed to find a problem in my `server_e2e_test.go`: requests are sent by reference, and I kept modifying the same instance, so in some cases a request was modified before it was being processed. That only affected the new test for stale nonce. All other tests were well sequenced.